lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionSourceCode.cpp

```diff
@@ -26,12 +26,95 @@ static const char *GetUserCodeStartMarker() {
 }
 static const char *GetUserCodeEndMarker() { return "\n/*__LLDB_USER_END__*/"; }
 
-static void WrapExpression(lldb_private::Stream &wrapped_stream,
-                           const char *orig_text, bool needs_object_ptr,
-                           bool static_method, bool is_class, bool weak_self,
-                           const EvaluateExpressionOptions &options,
-                           llvm::StringRef os_version,
-                           uint32_t &first_body_line) {
+
+struct CallsAndArgs {
+  std::string lldb_user_expr;
+  std::string lldb_trampoline;
+  std::string lldb_sink;
+  std::string lldb_call;
+};
+
+/// Constructs the signatures for the expression evaluation functions based on
+/// the metadata variables in scope.
+/// For every outermost metadata pointer in scope ($τ_0_0, $τ_0_1, etc), we want
+/// to generate:
+///
+/// - A $__lldb_user_expr signature that takes in that many metadata pointers:
+///
+/// func $__lldb_user_expr<T0, T1, ..., Tn>
+///     (_ $__lldb_arg: UnsafeMutablePointer<(T0, T1, ..., Tn)>)
+///
+/// - A $__lldb_trampoline signature like the above, but that also takes in a
+/// pointer to self:
+///
+/// func $__lldb_trampoline<T0, T1, ..., Tn>
+///      (_ $__lldb_arg: UnsafeMutablePointer<(T0, T1, ..., Tn)>,
+///       _ $__lldb_injected_self: inout $__lldb_context)
+///
+/// - A $__lldb_sink signature that matches the number of parameters of the
+/// trampoline:
+///
+/// func $__lldb_sink(_ $__lldb_arg : UnsafeMutablePointer<Any>,
+///                   _: $__lldb_builtin_ptr_t, // the self variable
+///                   _: $__lldb_builtin_ptr_t, // T0
+///                   _: $__lldb_builtin_ptr_t, // T1
+///                   ...,
+///                   _: $__lldb_builtin_ptr_t) // Tn
+///
+/// - And a matching call to the sink function:
+///
+/// lldb_sink($__lldb_arg, $__lldb_injected_self, $τ_0_0, $τ_0_1, ..., $τ_0_n)
+static CallsAndArgs MakeGenericSignaturesAndCalls(
+    llvm::ArrayRef<SwiftASTManipulator::VariableInfo> local_variables) {
+  llvm::SmallVector<SwiftASTManipulator::VariableInfo> metadata_variables;
+  for (auto &var : local_variables)
+    if (var.IsOutermostMetadataPointer())
+      metadata_variables.push_back(var);
+
+  StreamString generic_param_list_stream;
+  for (size_t i = 0; i < metadata_variables.size(); ++i)
+    generic_param_list_stream << "T" << std::to_string(i) << ", ";
```

Comment:
I don't think the `to_string()` is necessary when you use the stream? At least llvm::raw_ostream supports the conversions.

Reply:
`StreamString` actually doesn't support the conversion, I updated the patch to use `llvm::raw_string_ostream`